modules/services/kanata.nix

```diff
@@ -0,0 +1,171 @@
+              caps (tap-hold 100 100 caps lctl))
+          '';
+          description = ''
+            Configuration other than `defcfg`.
+
+            ${upstreamDoc}
+          '';
+        };
+        extraDefCfg = lib.mkOption {
+          type = lib.types.lines;
+          default = "";
+          example = "danger-enable-cmd yes";
+          description = ''
+            Configuration of `defcfg` other than `macos-dev-names-include` (generated
+            from the devices option).
+
+            ${upstreamDoc}
+          '';
+        };
+        configFile = lib.mkOption {
+          type = lib.types.path;
+          default = mkConfig name config;
+          defaultText = "A config file generated by values from other kanata module options.";
+          description = ''
+            The config file.
+
+            By default, it is generated by values from other kanata
+            module options.
+
+            You can also set it to your own full config file which
+            overrides all other kanata module options.  ${upstreamDoc}
+          '';
+        };
+        extraArgs = lib.mkOption {
+          type = lib.types.listOf lib.types.str;
+          default = [ ];
+          description = "Extra command line arguments passed to kanata.";
+        };
+        port = lib.mkOption {
+          type = lib.types.nullOr lib.types.port;
+          default = null;
+          example = 6666;
+          description = ''
+            Port to run the TCP server on. `null` will not run the server.
+          '';
+        };
+      };
+    };
+
+  mkName = name: "kanata-${name}";
+
+  mkDevices =
+    devices:
+    let
+      devicesString = lib.pipe devices [
+        (map (device: "\"" + device + "\""))
+        (lib.concatStringsSep " ")
+      ];
+    in
+    lib.optionalString ((lib.length devices) > 0) "macos-dev-names-include (${devicesString})";
+
+  mkConfig =
+    name: keyboard:
+    pkgs.writeTextFile {
+      name = "${mkName name}-config.kdb";
+      text = ''
+        (defcfg
+          ${keyboard.extraDefCfg}
+          ${mkDevices keyboard.devices})
+
+        ${keyboard.config}
+      '';
+      # Only the config file generated by this module is checked.  A
+      # user-provided one is not checked because it may not be available
+      # at build time.  I think this is a good balance between module
+      # complexity and functionality.
+      checkPhase = ''
+        ${lib.getExe cfg.package} --cfg "$target" --check --debug
+      '';
+    };
+
+  mkService =
+    name: keyboard:
+    lib.nameValuePair (mkName name) {
+      command = ''
+        ${lib.getExe cfg.package} \
+           --cfg ${keyboard.configFile} \
+           ${lib.optionalString (keyboard.port != null) "--port ${toString keyboard.port} \\"}
+           ${lib.optionalString (keyboard.extraArgs != "") builtins.concatStringsSep " " keyboard.extraArgs}
+      '';
+      serviceConfig = {
```

Comment:
path should also be passed in from an extraPackages option as commands will become useless unless in /sbin etc